src/OpenIddict.Core/OpenIddictBuilder.cs

```diff
@@ -1,19 +1,404 @@
+            if (!contract.IsAssignableFrom(typeof(TStore))) {
+                throw new InvalidOperationException("Custom stores must implement IOpenIddictAuthorizationStore.");
+            }
+
+            Services.Replace(ServiceDescriptor.Scoped(contract, typeof(TStore)));
+
+            return this;
+        }
+
+        /// <summary>
+        /// Replaces the default token manager by a custom manager.
+        /// </summary>
+        /// <typeparam name="TManager">The type of the custom manager.</typeparam>
+        /// <returns>The <see cref="OpenIddictBuilder"/>.</returns>
+        public virtual OpenIddictBuilder ReplaceTokenManager<TManager>() {
+            var contract = typeof(OpenIddictTokenManager<,>).MakeGenericType(TokenType, UserType);
+            if (!contract.IsAssignableFrom(typeof(TManager))) {
+                throw new InvalidOperationException("Custom managers must be derived from OpenIddictTokenManager.");
+            }
+
+            Services.Replace(ServiceDescriptor.Scoped(contract, typeof(TManager)));
+
+            return this;
+        }
+
+        /// <summary>
+        /// Replaces the default token store by a custom store.
+        /// </summary>
+        /// <typeparam name="TStore">The type of the custom store.</typeparam>
+        /// <returns>The <see cref="OpenIddictBuilder"/>.</returns>
+        public virtual OpenIddictBuilder ReplaceTokenStore<TStore>() {
+            var contract = typeof(IOpenIddictTokenStore<>).MakeGenericType(TokenType);
+            if (!contract.IsAssignableFrom(typeof(TStore))) {
+                throw new InvalidOperationException("Custom stores must implement IOpenIddictTokenStore.");
+            }
+
+            Services.Replace(ServiceDescriptor.Scoped(contract, typeof(TStore)));
+
+            return this;
+        }
+
+        /// <summary>
+        /// Registers a new OpenIddict module. If a module with the same name already
+        /// exists, the new instance is ignored and this extension has no effect.
+        /// </summary>
+        /// <param name="name">The name of the OpenIddict module.</param>
+        /// <param name="position">The relative position of the OpenIddict module in the ASP.NET Core pipeline.</param>
+        /// <param name="registration">The delegate used to register the module in the ASP.NET Core pipeline.</param>
+        /// <returns>The<see cref="OpenIddictBuilder"/>.</returns>
+        public virtual OpenIddictBuilder AddModule(
+            [NotNull] string name, int position,
+            [NotNull] Action<IApplicationBuilder> registration) {
+            if (string.IsNullOrEmpty(name)) {
+                throw new ArgumentNullException(nameof(name));
+            }
+
+            if (registration == null) {
+                throw new ArgumentNullException(nameof(registration));
+            }
+
+            return Configure(options => {
+                if (options.Modules.Any(module => module.Name == name)) {
```

Comment:
Unfortunately, there's no way to retrieve an `ILogger` instance at this stage (this method is called from `ConfigureServices()`).